drake/solvers/trajectoryOptimization/PseudoSpectralMethodTrajOpt.m

```diff
@@ -0,0 +1,221 @@
+      dxdotdu_f=(repmat(dxdotdu_all_knots,1,N)).*kron(eye(N),ones(nX,nU))*scale;
+      % d(xdot_p)/dx
+      dxdotdx_p=kron(D,speye(nX)); 
+
+      % difference between the dynamic and the psm fitting curve
+      g=[xdot_p]-[xdot_f];
+      % gradient of such difference
+      d_p=[sparse(N*nX,1),dxdotdx_p,sparse(nX*N,nU*N)]; % [dxdot_p/dh1, dxdot_p/dx,dxdot_p/du]
+      d_f=[dxdotdh1_f,dxdotdx_f,dxdotdu_f]; % dynamics gradient 
+      dg=d_p-d_f;      
+    end
+    
+    
+    function data = dynamics_data(obj,x,u)
+      [data.xdot,data.dxdot] = obj.plant.dynamics(0,x,u);
+    end
+    
+    function obj = addRunningCost(obj,running_cost_function)
+      % Adds an integrated cost to all time steps, which is
+      % numerical implementation specific 
+      % this cost is assumed to be time-invariant
+      % @param running_cost_function a function handle
+      %  of the form running_cost_function(dt,x,u)
+
+      nX = obj.plant.getNumStates();
+      nU = obj.plant.getNumInputs();
+      % the weight vector
+      w=obj.w;
+
+      % add running cost at the first node
+      running_cost_end = FunctionHandleObjective(nX+nU,@(x,u) obj.running_fun_end(running_cost_function,x,u));
+      obj = obj.addCost(running_cost_end,{obj.x_inds(:,1);obj.u_inds(:,1)});   
+      % add running cost at nodes 2 to N
+      for i=2:obj.N,
+        running_cost_mid = FunctionHandleObjective(1+nX+nU,@(h,x,u) obj.running_fun_mid(running_cost_function,h,x,u,w(i)));
+        obj = obj.addCost(running_cost_mid,{obj.h_inds(i-1);obj.x_inds(:,i);obj.u_inds(:,i)});
+      end
+    end
+    function [utraj,xtraj] = reconstructInputTrajectory(obj,z)
+      % Interpolate all knot points to reconstruct a trajectory using
+      % Lagrange interpolation
+      nU = obj.plant.getNumInputs();
+      u = reshape(z(obj.u_inds),[],obj.N);
+      % it is tempting to interpolate at the physical time domain nodes as it is the most convenient, but to minimize the effect of Runge phenomenon, 
+      % do the Lagrange interpolation at tau (Gauss quadratures). The shifting will be taken care of in LagrangeInterpolation method.
+      % note that, while this remedy does minimize the effect, it unfortunately doesn't gurantee the Runge phenomenon won't happen. Polyfit call will give warning
+      % when Runge phenomenon is unavoidable.
+      utraj=PPTrajectory(LagrangeInterpolation(obj.tau,u,nU));
+      % scale from [0,2] to the physical time domain
+      utraj = utraj.scaleTime(sum(z(obj.h_inds))/2);
+      utraj = utraj.setOutputFrame(obj.plant.getInputFrame);
+    end
+    
+    function xtraj = reconstructStateTrajectory(obj,z)
+      % Interpolate all knot points to reconstruct a trajectory using
+      % Lagrange interpolation
+      nX = obj.plant.getNumStates();
+      x = reshape(z(obj.x_inds),[],obj.N);
+      % similar reason as in reconstructInputTrajectory() for the choice of interpolating times
+      xtraj=PPTrajectory(LagrangeInterpolation(obj.tau,x,nX));
```

Comment:
Does this reconstruct Legendre polynomial only? How about Chebyshev polynomials?